crates/red_knot_test/src/assertion.rs

```diff
@@ -0,0 +1,621 @@
+    inner: std::iter::Peekable<AssertionWithRangeIterator<'a>>,
+}
+
+impl<'a> Iterator for LineAssertionsIterator<'a> {
+    type Item = LineAssertions<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let file = self.file_assertions;
+        let ranged_assertion = self.inner.next()?;
+        let mut collector = AssertionVec::new();
+        let mut line = file.line_number(&ranged_assertion);
+        // Collect all own-line comments on consecutive lines; these all apply to the same line of
+        // code. For example:
+        //
+        //     # Error: [unbound-name]
+        //     # Type: Unbound
+        //     reveal_type(x)
+        //
+        if file.is_own_line(&ranged_assertion) {
+            collector.push(ranged_assertion.into());
+            let mut only_own_line = true;
+            while let Some(ranged_assertion) = self.inner.peek() {
+                let next_line = line.saturating_add(1);
+                if file.line_number(ranged_assertion) == next_line {
+                    if !file.is_own_line(ranged_assertion) {
+                        only_own_line = false;
+                    }
+                    line = next_line;
+                    collector.push(self.inner.next().unwrap().into());
+                    // If we see an end-of-line comment, it has to be the end of the stack,
+                    // otherwise we'd botch this case, attributing all three errors to the `bar`
+                    // line:
+                    //
+                    //     # Error:
+                    //     foo  # Error:
+                    //     bar  # Error:
+                    if !only_own_line {
+                        break;
+                    }
+                } else {
+                    break;
+                }
+            }
+            if only_own_line {
+                // The collected comments apply to the _next_ line in the code.
+                line = line.saturating_add(1);
+            }
+        } else {
+            // We have a line-trailing comment; it applies to its own line, and is not grouped.
+            collector.push(ranged_assertion.into());
+        }
+        Some(LineAssertions {
+            line,
+            assertions: collector,
+        })
+    }
+}
+
+impl std::iter::FusedIterator for LineAssertionsIterator<'_> {}
+
+/// One or more assertions referring to the same line of code.
+#[derive(Debug)]
+pub(crate) struct LineAssertions<'a> {
+    /// The line these assertions refer to.
+    ///
+    /// Not necessarily the same line the assertion comment is located on; for an own-line comment,
+    /// it's the next non-assertion line.
+    pub(crate) line: OneIndexed,
+
+    /// The assertions referring to this line.
+    pub(crate) assertions: AssertionVec<'a>,
+}
+
+impl<'a, 'b> From<&'a LineAssertions<'b>> for &'a [Assertion<'b>] {
+    fn from(value: &'a LineAssertions<'b>) -> Self {
+        &value.assertions[..]
+    }
+}
```

Comment:
I think some of your `From` implementations would be better as `Deref`s. It would allow you to simplify a bunch of other signatures and remove a bunch of `.into()` calls. All tests pass with this diff:

<details>

```diff
diff --git a/crates/red_knot_test/src/assertion.rs b/crates/red_knot_test/src/assertion.rs
index f55291326..a7b0801ee 100644
--- a/crates/red_knot_test/src/assertion.rs
+++ b/crates/red_knot_test/src/assertion.rs
@@ -219,9 +219,11 @@ pub(crate) struct LineAssertions<'a> {
     pub(crate) assertions: AssertionVec<'a>,
 }
 
-impl<'a, 'b> From<&'a LineAssertions<'b>> for &'a [Assertion<'b>] {
-    fn from(value: &'a LineAssertions<'b>) -> Self {
-        &value.assertions[..]
+impl<'a> Deref for LineAssertions<'a> {
+    type Target = [Assertion<'a>];
+
+    fn deref(&self) -> &Self::Target {
+        &self.assertions
     }
 }
 
diff --git a/crates/red_knot_test/src/diagnostic.rs b/crates/red_knot_test/src/diagnostic.rs
index 7c5f940e6..25a28fea9 100644
--- a/crates/red_knot_test/src/diagnostic.rs
+++ b/crates/red_knot_test/src/diagnostic.rs
@@ -1,6 +1,8 @@
 //! Sort and group diagnostics by line number, so they can be correlated with assertions.
 //!
 //! We don't assume that we will get the diagnostics in source order.
+use std::ops::Deref;
+
 use ruff_source_file::{LineIndex, OneIndexed};
 use ruff_text_size::Ranged;
 use smallvec::SmallVec;
@@ -76,9 +78,11 @@ pub(crate) struct LineDiagnostics<T> {
     pub(crate) diagnostics: DiagnosticVec<T>,
 }
 
-impl<'a, T> From<&'a LineDiagnostics<T>> for &'a [T] {
-    fn from(value: &'a LineDiagnostics<T>) -> Self {
-        &value.diagnostics[..]
+impl<T> Deref for LineDiagnostics<T> {
+    type Target = [T];
+
+    fn deref(&self) -> &Self::Target {
+        &self.diagnostics
     }
 }
 
diff --git a/crates/red_knot_test/src/matcher.rs b/crates/red_knot_test/src/matcher.rs
index 899268bcc..2e3849012 100644
--- a/crates/red_knot_test/src/matcher.rs
+++ b/crates/red_knot_test/src/matcher.rs
@@ -131,8 +131,8 @@ impl Unmatched for Assertion<'_> {
     }
 }
 
-fn unmatched<'a, T: Unmatched + 'a>(unmatched: impl Into<&'a [T]>) -> Vec<String> {
-    unmatched.into().iter().map(Unmatched::unmatched).collect()
+fn unmatched<'a, T: Unmatched + 'a>(unmatched: &'a [T]) -> Vec<String> {
+    unmatched.iter().map(Unmatched::unmatched).collect()
 }
 
 struct Matcher {
@@ -153,15 +153,15 @@ impl Matcher {
     /// Return vector of [`Unmatched`] for any unmatched diagnostics or assertions.
     fn match_line<'a, 'b, T: Diagnostic + 'a>(
         &self,
-        diagnostics: impl Into<&'a [T]>,
-        assertions: impl Into<&'a [Assertion<'b>]>,
+        diagnostics: &'a [T],
+        assertions: &'a [Assertion<'b>],
     ) -> Result<(), Vec<String>>
     where
         'b: 'a,
     {
         let mut failures = vec![];
-        let mut unmatched = diagnostics.into().iter().collect::<Vec<_>>();
-        for assertion in assertions.into() {
+        let mut unmatched: Vec<&T> = diagnostics.iter().collect();
+        for assertion in assertions {
             if !self.matches(assertion, &mut unmatched) {
                 failures.push(assertion.unmatched());
```

</details>

Reply:
(If you don't like `Deref`s, then I'd use an `AsRef` implementation or an `as_slice` custom method like Micha suggested, rather than `From`.)

Reply:
Yes, that's definitely an improvement, thank you!